### PR TITLE
 Adopt Servo's Task Cluster integration

### PR DESCRIPTION
**Pull request by ncalexan**
There are many benefits to cloning Servo's Task Cluster integration:
- there's first-class support for generating Docker images in automation, which we don't do at all right now
- there's some support for flowing dependencies through task graphs (although it's not polished), which we will need very shortly for #263 
- the Python library is simple and well documented, and we can potentially collaborate to build a bigger ecosystem than just our repository
- it's built around [Task Cluster's v1 format](https://docs.taskcluster.net/docs/reference/integrations/taskcluster-github/docs/taskcluster-yml-v1), so we can pay down some debt along the way

/cc @SimonSapin I'm going to see if this is a good approach for https://github.com/lissyx/taskcluster-github-decision/issues/5.  Just a heads up!

### CODE REVIEW

**Comment by SimonSapin** [comment]
Cool, let me know if you have any question about the code or setup. There’s a `README.md` file next to where you found `decisionlib.py`.

The Python code should be mostly independent of `.taskcluster.yml` v1 v.s. v0. That yaml file only starts one decision task with some environment variables.

**Comment by ncalexan** [comment]
@SimonSapin I'm trying to use your `decisionlib`, and I'm getting odd failures in the Docker image building bits, which I've not (knowingly) altered.  It seems reproducible.  In, for example, https://tools.taskcluster.net/groups/UVLGdXn4T5OQfFYzkxDINQ/tasks/P9fWCKaqQuOAtJsHgMoqMg/runs/0/logs/public%2Flogs%2Flive.log, you can see that the image builds but then we can't save it:
```
+ docker save taskcluster-built
time="2018-10-24T02:06:27Z" level=fatal msg="Get http:///var/run/docker.sock/v1.18/images/taskcluster-built/get: EOF. Are you trying to connect to a TLS-enabled daemon without TLS?" 
```
The internet suggests to me that message means that Docker's daemon isn't running, suggesting that the Docker-in-Docker configuration isn't working.  I haven't changed the bootstrap image (knowingly).

Any thoughts?

**Comment by ncalexan** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?format=default&id=1390078 suggests this might be a very opaque way of complaining about our image size (which is, frankly, huge -- we have the Android SDK and NDK _and_ the Rust toolchains for three architectures, and the C/C++ toolchains, and Gradle).

**Comment by SimonSapin** [comment]
I haven’t seen that error before. To see if it’s indeed due to the image size, try building a smaller image?

FWIW in Servo we have rustup.rs and C/C++/Java toolchains in the image, but Rust toolchains and Android NDK/SDKs and Gradle-related things are downloaded at the start of each task (that need them). We do cache `~/.rustup` and `~/.gradle` on workers, though.

https://github.com/servo/servo/blob/b19f9d9c5b6360d1d05ac3b2b7b3b5d78b0bdfa9/python/servo/bootstrap_commands.py#L79-L220

https://docs.taskcluster.net/docs/reference/workers/docker-worker/docs/caches

**Comment by SimonSapin** [comment]
(Unsubscribing to not keep getting email for every commit, feel free to mention me again to discuss anything.)

**Comment by ncalexan** [comment]
Sorry that this is pretty big, but it does a lot of things.

File: automation/taskcluster/decision_task.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        # https://docs.taskcluster.net/docs/reference/workers/docker-worker/docs/caches
+        .with_scopes("docker-worker:cache:application-services-*")
+        .with_caches(**{
+            "application-services-cargo-registry": "/root/.cargo/registry",
+            "application-services-cargo-git": "/root/.cargo/git",
+            "application-services-sccache": "/root/.cache/sccache",
+            "application-services-gradle": "/root/.gradle",
+            # After we get the docker-in-docker image building working, we can
+            # do this instead of baking Rust into our images.
+            # "application-services-rustup": "/root/.rustup",
```
Comment (thomcc) [inline automation/taskcluster/decision_task.py]: Can you file an issue for this?

File: .taskcluster.yml [hunk 2]
```diff
@@ -2,132 +2,117 @@
...
+          - "queue:create-task:highest:aws-provisioner-v1/application-services-r"
+
+          # So that we can cache task outputs for re-use.
+          - "queue:route:index.project.application-services.*"
+
+          # So that we can re-use Gradle/Cargo/sccache bits between tasks.
+          - "docker-worker:cache:application-services-*"
+
+        payload:
+          maxRunTime: {$eval: '20 * 60'}
```
Comment (thomcc) [inline .taskcluster.yml]: Weird. Is there a reason to specify this separately?

File: automation/taskcluster/decision_task.py (hunk 1 above)
Comment (ncalexan) [reply automation/taskcluster/decision_task.py]: This is tracked by [Bug 1501720](https://bugzilla.mozilla.org/show_bug.cgi?id=1501720), but I've also filed https://github.com/mozilla/application-services/issues/339.

File: .taskcluster.yml (hunk 2 above)
Comment (ncalexan) [reply .taskcluster.yml]: The repetition of blocks based on the different events?  Not really, I just couldn't be bothered to smooth it out.  Filed https://github.com/mozilla/application-services/issues/340 to improve that.

Comment (thomcc) [reply .taskcluster.yml]: I more meant using some sort of evaluation syntax to compute 20 * 60

**Pull request closed without merging**